### PR TITLE
[Code Together]Added matrix form to Grover Operation

### PR DESCRIPTION
**Pull request by tgag17**
**Context:**
Added matrix form to the Grover Operation

**Description of the Change:**
Added the Grover Diffusion Operator which is customized i.e. generates the matrix for a n number of wires

**Benefits:**
Allows more efficient use of the template on statevector simulators.

**Related GitHub Issues:**
Issue #1514

### CODE REVIEW

**Comment by tgag17** [comment]
Really sorry for these typos

**Comment by tgag17** [comment]
I had a question about the formatting.
After doing black -l 100 pennylane a huge number of files were modified along with the ones in which I made the changes. Any suggestion about what should be done?

**Comment by antalszava** [comment]
Hi @tgag17, was the `black -l 100 pennylane` command executed from the external `pennylane` folder? There are two `pennylane` folders, one that contains the entire repository and the other that contains the source code of the project. We'd like to apply `black` on the latter one (along with the `tests` folder too).

**Comment by tgag17** [comment]
It was executed inside the external `pennylane` folder.

File: tests/templates/test_subroutines/test_grover.py [hunk 1]
```diff
@@ -118,3 +118,19 @@ def circ():
     probs = circ()
 
     assert np.argmax(probs) == len(probs) - 1
+
+
+@pytest.mark.parametrize("num", 2)
```
Comment (antalszava) [inline tests/templates/test_subroutines/test_grover.py:123]:
```suggestion
    @pytest.mark.parametrize("num", 2)
```

**Comment by antalszava** [comment]
Hi @tgag17, there seems to be a conflict in the changelog. This can be resolved by pulling master (`git checkout master` followed by `git pull`), switching to the feature branch (`git checkout add_reseterror_support`) and then merging `master` (`git merge master`). The culprit seems to be the changelog file. We'd like to have all the entries kept there, for some reason `git` could not resolve the addition there well.

File: pennylane/templates/subroutines/grover.py [hunk 2]
```diff
@@ -127,3 +129,21 @@ def expand(self):
                 Hadamard(wire)
 
         return tape
+
+    @classmethod
+    def _matrix(cls, *params):
+        num = cls.num_wires
```
Comment (antalszava) [inline pennylane/templates/subroutines/grover.py:135]: Hi @tgag17, slightly surprised that we'd get the correct number of wires here this way. :thinking: 

```python
import pennylane as qml

qml.templates.GroverOperator.num_wires
```
gives me
```
<WiresEnum.AnyWires: -1>
```

which is an internal placeholder we use to represent ops/templates that can take any number of wires. However, at this time, we'd already like to know the exact number of wires to use.

Maybe it's best to pass the number of wires to `_matrix` by [redefining the `matrix` method, see how it's done for QFT here](https://github.com/PennyLaneAI/pennylane/blob/742f2ff067d969a42e380968f36456c995dee2c9/pennylane/ops/qubit/non_parametric_ops.py#L1039).

**Comment by antalszava** [comment]
Hi @tgag17, it looks that another contribution came in that got accepted, so we'll close this PR. However, don't worry, you were close to completion here and we'll definitely give out digital swag for this work. :slightly_smiling_face: :tada: Also see that you have another PR being in the review round! :rocket:

**Comment by tgag17** [comment]
Yes! Thank you for all your help and the digital swag🎉 Looking forward to completing the other PR!

**Pull request closed without merging**